runtime/download.go

```diff
@@ -0,0 +1,123 @@
+package runtime
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"regexp"
+
+	"github.com/yourbase/yb/plumbing/log"
+	"go4.org/xdgdir"
+)
+
+const defaultCacheDir = "/tmp/yourbase/cache"
+
+func localCacheDir() string {
+	if cacheDir, exists := os.LookupEnv("YB_CACHE_DIR"); exists {
+		return cacheDir
+	}
+	// Tries to find a XDG Cache dir
+	if cacheDir := xdgdir.Cache.Path(); cacheDir != "" {
+		return filepath.Join(cacheDir, "yourbase")
+	}
+	return defaultCacheDir
+}
+
+func cacheFilenameForURL(url string) (string, error) {
+	reg, err := regexp.Compile("[^a-zA-Z0-9.]+")
+	if err != nil {
+		return "", fmt.Errorf("Can't compile regex: %v", err)
+	}
+	fileName := reg.ReplaceAllString(url, "")
+	return fileName, nil
+}
+
+func downloadFileWithCache(ctx context.Context, url string) (string, error) {
+	cacheDir := localCacheDir()
+
+	filename, err := cacheFilenameForURL(url)
+	if err != nil {
+		return "", err
+	}
+
+	if err := os.MkdirAll(cacheDir, 0700); err != nil {
+		return "", fmt.Errorf("creating dir %s: %v", cacheDir, err)
+	}
+
+	cacheFilename := filepath.Join(cacheDir, filename)
+	log.Infof("Downloading %s to cache as %s", url, cacheFilename)
+
+	// Exists, don't re-download
+	if fi, err := os.Stat(cacheFilename); !os.IsNotExist(err) && fi != nil {
+		// Cancellable request
+		req, err := http.NewRequestWithContext(ctx, "HEAD", url, nil)
+		if err != nil {
+			return "", err
+		}
+
+		// try HEAD'ing the URL and comparing to local file
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil {
+			return "", fmt.Errorf("fetching %s: %v", url, err)
+		}
```

Comment:
`resp.Body` needs to be closed, as per https://pkg.go.dev/net/http?tab=doc#Response.Body

```suggestion
		}
		resp.Body.Close()
```